Add example automation scripts.

craigyam opened the pull request:
- Add startup/cleanup and bookinfo demo scripts.
- Add example targets to the Makefile.
- Add HelloWorld test automation changes
- Add scripts to setup kubernetes environment for testing.
- Add make examples to the before_deploy. Also, check the branch
name 'example_automation' as that's the current branch name for
this work.
- Need to install the dependencies (docker-compose, a8ctl etc)
when doing the deploy since make examples needs them.
- Automate the helloworld demo.
- Add traffic-step and traffic-abort and update beta cli binary.
- Tag images with the branch version number and support the cli
recipe-run command.
- Push bookinfo examples to dockerhub.
- Dump debug in bookinfo script after failures.
- Reset the rules on failures to see if that resolves the issues
with the nginx service reporting as down.
- Add additional retries to the remaining testcases.
- Add k8s commands when dump debug info.
- Seeing other failures during some of the other testcases that
will probably be fixed by retries.
- Use kubectl commands for k8s demo runs when dumping debug data.
- Add the start/stop makefile targets to set the version. Fix the
version variable name in the push to dockerhub script.
- Add make targets to stop/start bookinfo.
- Fix version tag for examples images.
- Move push to after_deploy in travis.yaml.
- The examples are running twice as the push script is considered
another deploy.  Move it to the after_deploy section.
Also, add stop/start makefile support for bookinfo.
- Call push script in after_deploy.
- Update push script and move versioned yamls to subdirs.
- Merge in examples updates to change to k8s registry.
- Update with elasticsearch changes.
- Build the cli before running the tests.
- Dockerize before running examples

craigyam committed:
Add example automation scripts.

- Add helloworld/bookinfo example automation scripts.
- Add scripts to start the control plane.

rshriram reviewed (changes requested):
@craigyam this patch is super huge. Can you break this up into smaller focussed commits that make things easier to review?

There is also an awful lot of duplication in the test scripts between testing and examples/test-scripts folder. Is there any way to deduplicate the stuff?

craigyam committed:
Update makefile to include examples targets.

- Add example targets to the Makefile.
- Update travis yaml to run make examples.
- Update dockerhub push script to push versioned examples images.

craigyam committed:
Automate helloworld docker examples demo.

- Add HelloWorld test automation changes

craigyam committed:
Automate helloworld Kubernetes examples demo.

craigyam committed:
Automate bookinfo docker examples demo.

craigyam committed:
Automate bookinfo Kubernetes examples demo.

craigyam committed:
Add expected output files and rules for examples automation.

Add the json files that describe the expected output from the
examples automated tests.  Add the json files with the rules that
need to be created while running the examples automated tests.

craigyam commented:
@rshriram @ijsnellf 
I have broken up the commit into smaller pieces so hopefully that makes reviewing a little easier.  I talked to @gicaceres about reusing code between examples and testing.  He's currently making a number of updates to testing so we thought we'd work together after he's done to see if there is a good way to share common code (in a separate pull request).  Sound good?

craigyam committed:
Update adapter env var name.

rshriram commented on examples/apps/bookinfo/build-services.sh:
```diff
@@ -16,35 +16,40 @@
 
+APP_VER=$1
+if [ -z $APP_VER ]; then
+	APP_VER="unknown"
```
we cannot build with unknown version. We should rather exit

rshriram commented on .travis.yml:
```diff
@@ -23,6 +24,9 @@ before_install:
+        wget http://stedolan.github.io/jq/download/linux64/jq;
+        chmod +x ./jq;
+        sudo mv jq /usr/local/bin;
       fi
```
I think we already have jq in the travis vm. Our existing tests use it IIRC. right?

rshriram commented on .travis.yml:
```diff
@@ -61,18 +66,9 @@ deploy:
-            tags: true
-            repo: amalgam8/amalgam8
-            condition: $TRAVIS_TAG =~ v[0-9]+\.[0-9]+\.[0-9]+
+    - ./scripts/push_dockerhub.sh
```
Why did you remove the check for version-tagged commit ?

rshriram commented on Makefile:
```diff
@@ -281,6 +282,77 @@ release.examples:
+
+examples.runhello: build.cli
+	@echo "--> running examples hello world"
+	@examples/test-scripts/docker/test-docker-helloworld.sh $(APP_VER_ABBR) || exit 1
```
You don't need `exit 1` in make file.. It should exit by default.

rshriram commented on examples/k8s-start-helloworld.sh:
```diff
@@ -0,0 +1,32 @@
+export A8_CONTROLLER_URL=http://localhost:31200
+export GATEWAY_URL=http://localhost:32000
+
+sed -e "s/{APP_VER}/$APP_VER/" $SCRIPTDIR/test-scripts/kubernetes/k8s-helloworld.yaml | kubectl create -f -
```
Is this file just for testing or for release? We don't need this file for release.

rshriram commented on examples/test-scripts/bookinfo.sh:
```diff
@@ -0,0 +1,491 @@
+#!/bin/bash
+#
```
reuse with testing/scripts ?

rshriram commented on examples/test-scripts/docker/cleanup-helloworld.sh:
```diff
@@ -0,0 +1,29 @@
+
+docker-compose -f $SCRIPTDIR/docker-helloworld.yaml kill
+docker-compose -f $SCRIPTDIR/docker-helloworld.yaml rm -f
+
```
why do we need this file?

rshriram commented on examples/test-scripts/docker/docker-bookinfo.yaml:
```diff
@@ -0,0 +1,98 @@
+
+details-v1:
+  image: amalgam8/a8-examples-bookinfo-details-sidecar-v1:${APP_VER}
+  environment:
```
Our test environment docker file is 99% identical. I suggest we export a bunch of env vars and use that to test actual example apps or test apps, using one single docker file.

rshriram commented on examples/test-scripts/docker/run-controlplane-docker.sh:
```diff
@@ -0,0 +1,79 @@
+else
+    echo "usage: $0 start|stop"
+    exit 1
+fi
```
Kill file. Please see comments on #463

rshriram commented on examples/test-scripts/docker/test-docker-bookinfo.sh:
```diff
@@ -0,0 +1,52 @@
+
+sleep 10
+
+exit $RC
```
Kill file. Please see comments on #463

rshriram commented on examples/test-scripts/gremlin-script/load-productpage.sh:
```diff
@@ -0,0 +1,6 @@
+# This script runs after the a8ctl recipe-run sets the rules, but before the logs
+# are checked.
+sleep 20
+curl -s -b 'foo=bar;user=jason;x' http://localhost:32000/productpage/productpage > /dev/null
```
kill. This is a duplicate.

rshriram commented on examples/test-scripts/helloworld.sh:
```diff
@@ -0,0 +1,243 @@
+    echo ""
+
+    if [ $v1_count -lt 70 ] || [  $v1_count -gt 80 ]; then
+        # Test failed, routing outside 75%/25%, try again
```
I am not sure you can rely on such accurate statistics, esp given that choice of versions is random. You might have to run it 1000 times.. Which would take forever.

rshriram commented on examples/test-scripts/kubernetes/cleanup-bookinfo.sh:
```diff
@@ -0,0 +1,27 @@
+
+sed -e "s/{APP_VER}/$APP_VER/" $SCRIPTDIR/k8s-bookinfo.yaml | kubectl delete -f - || echo "Probably already down"
+sleep 2
+$SCRIPTDIR/run-controlplane-kubernetes.sh stop
```
Kill file. Please see comments on #463

rshriram commented on examples/test-scripts/kubernetes/cleanup-helloworld.sh:
```diff
@@ -0,0 +1,27 @@
+
+sed -e "s/{APP_VER}/$APP_VER/" $SCRIPTDIR/k8s-helloworld.yaml | kubectl delete -f - || echo "Probably already down"
+sleep 2
+$SCRIPTDIR/run-controlplane-kubernetes.sh stop
```
kill file

rshriram replied:
Move the one liner code into the parent file.

rshriram commented on examples/test-scripts/kubernetes/install-kubernetes.sh:
```diff
@@ -0,0 +1,51 @@
+# Install kubernetes CLI
+type kubectl >/dev/null 2>&1 || install_kubectl
+echo "Waiting for K8S to initialize.."
+sleep 30
```
reuse from testing

rshriram commented on examples/test-scripts/kubernetes/k8s-bookinfo.yaml:
```diff
@@ -0,0 +1,312 @@
+        - name: A8_ELASTICSEARCH_SERVER
+          value: $(LOGSERVER_SERVICE_HOST):9200
+        - name: A8_SERVICE
+          value: productpage:version=v1
```
same comment as docker file.

rshriram commented on examples/test-scripts/kubernetes/k8s-helloworld.yaml:
```diff
@@ -0,0 +1,117 @@
+        - containerPort: 5000
+        env:
+        - name: A8_SERVICE
+          value: helloworld:version=v2
```
If this is the only difference in all these files, why can't we use bunch of environment variables ?

rshriram commented on examples/test-scripts/kubernetes/run-controlplane-kubernetes.sh:
```diff
@@ -0,0 +1,75 @@
+else
+    echo "usage: $0 start|stop"
+    exit 1
+fi
```
kill file.. see #463

rshriram commented on scripts/push_dockerhub.sh:
```diff
@@ -45,6 +45,25 @@ patch_tag="$major.$minor.$patch"
 major_tag="$major"
 
+echo "Tag: $tag"
+
```
Shouldn't we follow the same convention as sidecar, controller and registry?

rshriram reviewed (changes requested):
@craigyam Sorry for the delayed review. There is a lot of duplication between this and the test infrastructure we currently have. IF we take this patch as is, it will get immensely complicated to maintain two sets of scripts that do exactly the same thing.

I suggest reusing the testing scripts in the testing/ folder. Add additional code there to test out the apps in examples/ instead of testing out the apps in testing/apps/. 

Once that is done, the only other change that you would have to tackle is handling JSON (testapps) vs html (existing examples). I think this should be an easy fix.

With regard to helloworld, I suggest doing some simple statistics (50/50) and testing it over 100, and seeing if each version gets atleast 40+ hits.
